crates/ironrdp-rdpdr/src/pdu/efs.rs

```diff
@@ -0,0 +1,229 @@
+    PAKID_CORE_DEVICELIST_ANNOUNCE = 0x4441,
+    PAKID_CORE_DEVICE_REPLY = 0x6472,
+    PAKID_CORE_DEVICE_IOREQUEST = 0x4952,
+    PAKID_CORE_DEVICE_IOCOMPLETION = 0x4943,
+    PAKID_CORE_SERVER_CAPABILITY = 0x5350,
+    PAKID_CORE_CLIENT_CAPABILITY = 0x4350,
+    PAKID_CORE_DEVICELIST_REMOVE = 0x444D,
+    PAKID_PRN_CACHE_DATA = 0x5043,
+    PAKID_CORE_USER_LOGGEDON = 0x554C,
+    PAKID_PRN_USING_XPS = 0x5543,
+}
+
+impl std::convert::TryFrom<u16> for PacketId {
+    type Error = PduError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match value {
+            0x496E => Ok(PacketId::PAKID_CORE_SERVER_ANNOUNCE),
+            0x4343 => Ok(PacketId::PAKID_CORE_CLIENTID_CONFIRM),
+            0x434E => Ok(PacketId::PAKID_CORE_CLIENT_NAME),
+            0x4441 => Ok(PacketId::PAKID_CORE_DEVICELIST_ANNOUNCE),
+            0x6472 => Ok(PacketId::PAKID_CORE_DEVICE_REPLY),
+            0x4952 => Ok(PacketId::PAKID_CORE_DEVICE_IOREQUEST),
+            0x4943 => Ok(PacketId::PAKID_CORE_DEVICE_IOCOMPLETION),
+            0x5350 => Ok(PacketId::PAKID_CORE_SERVER_CAPABILITY),
+            0x4350 => Ok(PacketId::PAKID_CORE_CLIENT_CAPABILITY),
+            0x444D => Ok(PacketId::PAKID_CORE_DEVICELIST_REMOVE),
+            0x5043 => Ok(PacketId::PAKID_PRN_CACHE_DATA),
+            0x554C => Ok(PacketId::PAKID_CORE_USER_LOGGEDON),
+            0x5543 => Ok(PacketId::PAKID_PRN_USING_XPS),
+            _ => Err(invalid_message_err!("try_from", "PacketId", "invalid value")),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum VersionAndIdPDUKind {
+    /// [2.2.2.2 Server Announce Request (DR_CORE_SERVER_ANNOUNCE_REQ)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/046047aa-62d8-49f9-bf16-7fe41880aaf4)
+    ServerAnnounceRequest,
+    /// [2.2.2.3 Client Announce Reply (DR_CORE_CLIENT_ANNOUNCE_RSP)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/d6fe6d1b-c145-4a6f-99aa-4fe3cdcea398)
+    ClientAnnounceReply,
+}
+
+#[derive(Debug)]
+/// VersionAndIdPDU is a fixed size structure representing multiple PDUs.
+/// The kind field is used to determine the actual PDU type.
+pub struct VersionAndIdPDU {
+    version_major: u16,
+    version_minor: u16,
+    pub client_id: u32,
+    kind: VersionAndIdPDUKind,
+}
+
+impl VersionAndIdPDU {
+    pub fn new(version_major: u16, version_minor: u16, client_id: u32, kind: VersionAndIdPDUKind) -> Self {
+        Self {
+            version_major,
+            version_minor,
+            client_id,
+            kind,
+        }
+    }
+
+    fn header(&self) -> PduResult<SharedHeader> {
+        match self.kind {
+            VersionAndIdPDUKind::ClientAnnounceReply => Ok(SharedHeader {
+                component: Component::RDPDR_CTYP_CORE,
+                packet_id: PacketId::PAKID_CORE_CLIENTID_CONFIRM,
+            }),
+            VersionAndIdPDUKind::ServerAnnounceRequest => Err(unexpected_message_type_err!("ServerAnnounceRequest", 0)),
```

Comment:
I’m a bit curious about this. I understand we can’t encode a `VersionAndIdPDU` when the kind is "Server Announce Request", is there a reason for that? In general this kind of logic should not live in the PDU itself.

Reply:
It's not that we actually don't know how to do build and encode that PDU, it was just representative of the fact that we're currently just building the client side code and so don't currently need to be able to do so.

However I suppose there's no harm in just going ahead and implementing this now for later use by IronRDP server: 

6d7af41bb314c4a0ae13db3a7b15303665c1307d

Reply:
Thank you!